File: C++/Laba3sem2v2/queue.cpp
#include "queue.h"
#include <cstring>
#include <fstream>
#include <iostream>

using namespace std;

void Queue::init() {
    front = nullptr; // Инициализация указателя на начало очереди.
    rear = nullptr;  // Инициализация указателя на конец очереди.
}

void Queue::enqueue(const string& value) {
    QueueNode* newNode = new QueueNode{value, nullptr}; // Создание нового узла.
    if (rear != nullptr) {
        rear->next = newNode; // Связывание нового узла с текущим последним узлом.
    }
    rear = newNode; // Обновление rear на новый узел.
    if (front == nullptr) {
        front = rear; // Установка front, если очередь была пустой.
    }
}

void Queue::dequeue() {
    if (front == nullptr) {
        return; // Проверка на пустую очередь.
    }
    QueueNode* temp = front; // Сохранение текущего первого узла для удаления.
    front = front->next; // Перемещение front на следующий узел.
    if (front == nullptr) {
        rear = nullptr; // Обновление rear, если очередь стала пустой.
    }
    delete temp; // Освобождение памяти удаляемого узла.
}

void Queue::print() {
    QueueNode* temp = front; // Начало с первого узла.
    while (temp != nullptr) {
        cout << temp->data << " "; // Вывод данных узла.
        temp = temp->next; // Переход к следующему узлу.
    }
    cout << endl; // Печать новой строки после вывода всех элементов.
}

void Queue::destroy() {
    while (front != nullptr) {
        dequeue(); // Освобождение памяти всех узлов.
    }
}

void Queue::loadFromFile(const string& fileName) {
    ifstream file(fileName); // Открытие файла для чтения.
    string value;
    while (file >> value) {
        enqueue(value); // Добавление значений из файла в очередь.
    }
    file.close(); // Закрытие файла.
}

void Queue::saveToFile(const string& fileName) {
    ofstream file(fileName); // Открытие файла для записи.
    QueueNode* temp = front; // Начало с первого узла.
    while (temp != nullptr) {
        file << temp->data << endl; // Запись данных узла в файл.
        temp = temp->next; // Переход к следующему узлу.
    }
    file.close(); // Закрытие файла.
}

void runQueue(int argc, char* argv[]) {
    Queue queue;
    queue.init(); // Инициализация очереди.

    string fileName;
    string query;

    // Обработка аргументов командной строки для получения имени файла и запроса.
    for (int i = 1; i < argc; i++) {
        if (strcmp(argv[i], "--file") == 0 && i + 1 < argc) {
            fileName = argv[i + 1];
            i++;
        } else if (strcmp(argv[i], "--query") == 0 && i + 1 < argc) {
            query = argv[i + 1];
            i++;
        }
    }

    queue.loadFromFile(fileName); // Загрузка данных из файла в очередь.

    string command;
    size_t pos = query.find(' '); // Поиск пробела для разделения команды и аргумента.
    if (pos != string::npos) {
        command = query.substr(0, pos); // Извлечение команды.
        query = query.substr(pos + 1); // Извлечение аргумента.
    } else {
        command = query; // Если пробела нет, вся строка - это команда.
    }

    // Выполнение команды в зависимости от введенного запроса.
    if (command == "QPUSH") {
        queue.enqueue(query); // Добавление элемента в очередь.
        queue.saveToFile(fileName); // Сохранение состояния очереди в файл.
    } else if (command == "QPOP") {
        queue.dequeue(); // Удаление элемента из очереди.
        queue.saveToFile(fileName); // Сохранение состояния очереди в файл.
    } else if (command == "QPRINT") {
        queue.print(); // Вывод элементов очереди.
    }

    queue.destroy(); // Освобождение памяти перед завершением программы.
}

File: sem3l3go/queue/queue.go
package queue

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

// QueueNode представляет узел очереди
type QueueNode struct {
	Data string
	Next *QueueNode
}

// Queue представляет очередь
type Queue struct {
	Front *QueueNode
	Rear  *QueueNode
}

// NewQueue создает новую очередь
func NewQueue() *Queue {
	return &Queue{
		Front: nil,
		Rear:  nil,
	}
}

// Enqueue добавляет элемент в конец очереди
func (q *Queue) Enqueue(value string) {
	newNode := &QueueNode{Data: value, Next: nil}
	if q.Rear != nil {
		q.Rear.Next = newNode
	}
	q.Rear = newNode
	if q.Front == nil {
		q.Front = q.Rear
	}
}

// Dequeue удаляет элемент из начала очереди
func (q *Queue) Dequeue() bool {
	if q.Front == nil {
		return false
	}
	q.Front = q.Front.Next
	if q.Front == nil {
		q.Rear = nil
	}
	return true
}

// Print выводит все элементы
func (q *Queue) Print() {
	temp := q.Front
	for temp != nil {
		fmt.Printf("%s ", temp.Data)
		temp = temp.Next
	}
	fmt.Println()
}

// Destroy очищает очередь
func (q *Queue) Destroy() {
	for q.Front != nil {
		q.Dequeue()
	}
}

// LoadFromFile загружает из файла
func (q *Queue) LoadFromFile(filename string) error {
	if filename == "" {
		return nil
	}
	file, err := os.Open(filename)
	if err != nil {
		return err
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		q.Enqueue(scanner.Text())
	}
	return scanner.Err()
}

// SaveToFile сохраняет в файл
func (q *Queue) SaveToFile(filename string) error {
	if filename == "" {
		return nil
	}
	file, err := os.Create(filename)
	if err != nil {
		return err
	}
	defer file.Close()

	temp := q.Front
	for temp != nil {
		if _, err := fmt.Fprintln(file, temp.Data); err != nil {
			return err
		}
		temp = temp.Next
	}
	return nil
}

// RunQueue выполняет команды над очередью
func RunQueue(args []string) {
	queue := NewQueue()
	var filename string
	var query string

	for i := 0; i < len(args); i++ {
		if args[i] == "--file" && i+1 < len(args) {
			filename = args[i+1]
			i++
		} else if args[i] == "--query" && i+1 < len(args) {
			query = args[i+1]
			i++
		}
	}

	if filename != "" {
		queue.LoadFromFile(filename)
	}

	parts := strings.SplitN(query, " ", 2)
	command := parts[0]
	arg := ""
	if len(parts) > 1 {
		arg = parts[1]
	}

	switch command {
	case "QPUSH":
		if arg != "" {
			queue.Enqueue(arg)
			if filename != "" {
				queue.SaveToFile(filename)
			}
		}
	case "QPOP":
		queue.Dequeue()
		if filename != "" {
			queue.SaveToFile(filename)
		}
	case "QPRINT":
		queue.Print()
	}

	queue.Destroy()
}
